reaper/reaper.go

```diff
@@ -17,141 +17,153 @@
-	err = json.Unmarshal(body, &listServicesResp)
-	if err != nil {
-		fatalError("Invalid GET /v2/services response JSON", err, fatal)
+	if errorsFound {
+		return errors.New("errors occurred whilst reaping")
 	}
 
-	serviceGuid := listServicesResp.Resources[0].Metadata.Guid
+	return nil
+}
 
-	bodyReader, statusCode, err = authClient.DoAuthenticatedGet(apiUrl+fmt.Sprintf("/v2/services/%s/service_plans?results-per-page=%d", serviceGuid, MAXIMUM_RESULTS_PER_PAGE), accessToken)
-	if err != nil {
-		fatalError("GET /v2/services/[service GUID]/service_plans failed", err, fatal)
-	}
-	if statusCode != http.StatusOK {
-		fatal(fmt.Sprintf("GET /v2/services/[service GUID]/service_plans failed: %d", statusCode))
-	}
+func (r *reaper) servicesWithName(serviceName string) <-chan cloudfoundry.Service {
+	output := make(chan cloudfoundry.Service)
 
-	if bodyReader == nil {
-		fatal("GET /v2/services/[service GUID]/service_plans response body missing")
-	}
-	body, err = ioutil.ReadAll(bodyReader)
-	if err != nil {
-		fatalError("Cannot read GET /v2/services/[service GUID]/service_plans response body: %s", err, fatal)
-	}
+	go func() {
+		defer close(output)
 
-	var listServicePlansResp ListServicePlansResp
-	err = json.Unmarshal(body, &listServicePlansResp)
-	if err != nil {
-		fatalError("Invalid GET /v2/services/[service GUID]/service_plans response JSON", err, fatal)
-	}
+		services, err := r.cf.GetServices(serviceName)
+		if err != nil {
+			r.errorChannel <- err
+			return
+		}
 
-	for _, sp := range listServicePlansResp.Resources {
-		if sp.Entity.Free {
-			bodyReader, statusCode, err = authClient.DoAuthenticatedGet(apiUrl+fmt.Sprintf("/v2/service_plans/%s/service_instances?results-per-page=%d", sp.Metadata.Guid, MAXIMUM_RESULTS_PER_PAGE), accessToken)
-			if err != nil {
-				fatalError("GET /v2/service_plans/[service plan GUID]/service_instances failed", err, fatal)
-			}
-			if statusCode != http.StatusOK {
-				fatal(fmt.Sprintf("GET /v2/service_plans/[service plan GUID]/service_instances: %d", statusCode))
-			}
+		if len(services) == 0 {
+			fmt.Fprintf(r.output, "No services of type '%s' found", serviceName)
+			return
+		}
 
-			if bodyReader == nil {
-				fatal("GET /v2/service_plans/[service plan GUID]/service_instances response body missing")
-			}
-			body, err = ioutil.ReadAll(bodyReader)
-			if err != nil {
-				fatalError("Cannot read GET /v2/service_plans/[service plan GUID]/service_instances response body: %s", err, fatal)
-			}
+		output <- services[0]
+	}()
 
-			var servicePlanListServiceInstancesResp ServicePlanListServiceInstancesResp
-			err = json.Unmarshal(body, &servicePlanListServiceInstancesResp)
-			if err != nil {
-				fatalError("Invalid GET /v2/service_plans/[service plan GUID]/service_instances response JSON", err, fatal)
-			}
+	return output
+}
+
+func (r *reaper) freeServicePlansOf(services <-chan cloudfoundry.Service) <-chan cloudfoundry.ServicePlan {
+	output := make(chan cloudfoundry.ServicePlan)
+
+	go func() {
```

Comment:
Can we get away with making this a synchronous call as it would simplify the code

Reply:
Everything was synchronous until I started on handling paginated responses from cf, at which point I introduced channels in the `cf.go` functions as an alternative to the pagination control leaking out into `reaper.go`. And then at that point, given that goroutines and channels make it easy, it seemed pretty straightforward just to [pipeline](https://blog.golang.org/pipelines) the whole thing on line 50 - but that approach does rely on each step running in a goroutine, otherwise it'll block. I didn't worry about it too much since we seem to be going Reactive in lots of other places!

It's certainly possible to make this synchronous, although with some effort now that I've gone down this route. The goroutines in reaper would be replaced with more nested loops, and it'd have to be made aware of service plan and service instance pages so it can request one at a time. I'm not sure it's worth it, but at the same time I'd rather change it than commit a maintenance headache. What do you reckon?

Reply:
Thats cool, I just wanted to check the rationale and it sounds like we should!